### PR TITLE
VGM and SPC Support (Port from Clementine)

### PR DESCRIPTION
**Pull request by Eoin-ONeill-Yokai**
### SPC and VGM Support

I've gone ahead and ported most of the functionality from Clementine over to Strawberry. For the most part, it was a pretty straight forward conversion, but there are some specific changes that should be reviewed (specifically the `TagReaderWorker` class changes and how the multiple tag parsers are currently implemented.) For more details, see #991 ...

There are still some things that would be nice-to-haves (specifically, I would like to add reading / writing at some point w/ APE tag support for SPC) but I think that simply having the functionality up to date with Clementine will be good enough for now.

Another note, there are a few warnings due to `unsigned char` casting -- which I would eventually like to address as well using better logic (preferably, cast the signed bytes to u32int_t immediately and then do the binary shifts) but I think this MR is a good start. 

---

### Lastly, it's probably worth making sure that all users are up-to-date on the current state of libGME (the playback library that gstreamer uses) as its included in most linux distributions... 

- SPC format is basically fully implemented and fully functional. Most SNES `.spc` format files should run **without any major** accuracy issues.

- VGM format is more complicated. For the most part, most SEGA Genesis songs will work. However, the VGM format is a wide umbrella and unfortunately not all system implementations are supported in the standard libgme that is supplied with linux distributions. Those feeling adventurous though can link to a custom GME implementation built from source, but it does introduce some stability issues. (If you're really curious about that last one, check out my repositories to find a fork of one of these custom GME implementations.) 

 
I think that VGM will improve eventually, but it's hard to find the right people involved who want to make this filetype work better on Linux through gstreamer -- In general we're talking about a niche-within-a-niche here. This is something I do want to work on though, I just wish my knowledge of gstreamer was better. Projects such as ValleyBell/libvgm give me some hope that this situation will improve eventually. 

Anyway, let me know if there's any changes that need to be made here. I hope this port is satisfactory, but I can make changes here-or-there to accommodate whatever design goals the :strawberry: Strawberry project has in mind.

### CODE REVIEW

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+  QFileInfo fileinfo(filename);
+  return GME::IsSupportedFormat(fileinfo);
+}
+
+void TagReaderGME::ReadFile(const QString &filename, spb::tagreader::SongMetadata *song) const {
+  QFileInfo fileinfo(filename);
+  GME::ReadFile(fileinfo, song);
+}
+
+bool TagReaderGME::SaveFile(const QString &filename, const spb::tagreader::SongMetadata &song) const {
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:261]:
```suggestion
bool TagReaderGME::SaveFile(const QString&, const spb::tagreader::SongMetadata&) const {
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  return false;
+}
+
+QByteArray TagReaderGME::LoadEmbeddedArt(const QString &filename) const {
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:265]:
```suggestion
QByteArray TagReaderGME::LoadEmbeddedArt(const QString&) const {
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+  return QByteArray();
+}
+
+bool TagReaderGME::SaveEmbeddedArt(const QString &filename, const QByteArray &data) {
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:269]:
```suggestion
bool TagReaderGME::SaveEmbeddedArt(const QString&, const QByteArray&) {
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+  return false;
+}
+
+bool TagReaderGME::SaveSongPlaycountToFile(const QString &filename, const spb::tagreader::SongMetadata &song) const {
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:273]:
```suggestion
bool TagReaderGME::SaveSongPlaycountToFile(const QString&, const spb::tagreader::SongMetadata&) const {
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+  return false;
+}
+
+bool TagReaderGME::SaveSongRatingToFile(const QString &filename, const spb::tagreader::SongMetadata &song) const {
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:277]:
```suggestion
bool TagReaderGME::SaveSongRatingToFile(const QString&, const spb::tagreader::SongMetadata&) const {
```

File: ext/libstrawberry-tagreader/gmereader.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
+#ifndef GMEREADER_H
+#define GMEREADER_H
+
+#include <QtCore>
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.h:4]: Replace this include with QByteArray, QString and QFileInfo.

File: ext/libstrawberry-tagreader/gmereader.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+#include <taglib/tstring.h>
+#include "tagreaderbase.h"
+
+#include "tagreadermessages.pb.h"
+
+class QFileInfo;
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.h:10]: Remove

File: ext/libstrawberry-tagreader/gmereader.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+class QByteArray;
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.h:11]: Remove

File: ext/libstrawberry-tagreader/gmereader.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+namespace GME {
+bool IsSupportedFormat(const QFileInfo& file_info);
+void ReadFile(const QFileInfo& file_info,
+              spb::tagreader::SongMetadata* song_info);
+
+namespace SPC {
+/* SPC SPEC:
+ * http://vspcplay.raphnet.net/spc_file_format.txt
+ */
+const int HAS_ID6_OFFSET = 0x23;
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.h:22]: constexpr for all these

File: ext/strawberry-tagreader/tagreaderworker.cpp [hunk 10]
```diff
@@ -34,36 +34,56 @@ void TagReaderWorker::MessageArrived(const spb::tagreader::Message &message) {
 
   spb::tagreader::Message reply;
 
+  bool success = HandleMessage(message, reply, &tag_reader_);
+  if (!success || message.has_read_file_request()) {
```
Comment (jonaski) [inline ext/strawberry-tagreader/tagreaderworker.cpp:38]:
```suggestion
  if (!success && message.has_read_file_request()) {
```

File: ext/strawberry-tagreader/tagreaderworker.h [hunk 11]
```diff
@@ -44,11 +46,16 @@ class TagReaderWorker : public AbstractMessageHandler<spb::tagreader::Message> {
...
+  // Handle message using specific TagReaderBase implementation. Returns true on successful message handle.
+  bool HandleMessage(const spb::tagreader::Message &message, spb::tagreader::Message &reply, TagReaderBase* reader);
+
 #if defined(USE_TAGLIB)
   TagReaderTagLib tag_reader_;
 #elif defined(USE_TAGPARSER)
   TagReaderTagParser tag_reader_;
 #endif
+
+  TagReaderGME tag_reader_gme;
```
Comment (jonaski) [inline ext/strawberry-tagreader/tagreaderworker.h:58]:
```suggestion
  TagReaderGME tag_reader_gme_;
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+  file.seek(file.pos() + 4);
+  QByteArray gd3_length_bytes = file.read(4);
+  quint32 gd3_length =
+      (unsigned char)gd3_length_bytes[0] | ((unsigned char)gd3_head[1] << 8) |
+      ((unsigned char)gd3_head[2] << 16) | ((unsigned char)gd3_head[3] << 24);
+
+  QByteArray gd3Data = file.read(gd3_length);
+  QTextStream fileTagStream(gd3Data, QIODevice::ReadOnly);
+  // Stored as 16 bit UTF string, two bytes per letter.
+  fileTagStream.setEncoding(QStringConverter::Utf16);
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:199]: This does not compile with Qt 5. Need to use a ifdef here and use `QTextStream::setCodec` with Qt 5.

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+#include <QtEndian>
+
+#include "core/logging.h"
+#include "core/timeconstants.h"
+#include "core/messagehandler.h"
+#include "tagreaderbase.h"
+#include "tagreadertaglib.h"
+
+bool GME::IsSupportedFormat(const QFileInfo& file_info) {
+  return file_info.exists() &&
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:20]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+          (file_info.completeSuffix().endsWith("spc") ||
+                file_info.completeSuffix().endsWith("vgm"));
+}
+
+void GME::ReadFile(const QFileInfo& file_info,
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:25]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+                   spb::tagreader::SongMetadata* song_info) {
+  if (file_info.completeSuffix().endsWith("spc"))
+    SPC::Read(file_info, song_info);
+  if (file_info.completeSuffix().endsWith("vgm"))
+    VGM::Read(file_info, song_info);
+}
+
+void GME::SPC::Read(const QFileInfo& file_info,
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:33]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+  file.seek(INTRO_LENGTH_OFFSET);
+  QByteArray length_bytes = file.read(INTRO_LENGTH_SIZE);
+  quint64 length_in_sec = 0;
+  if (length_bytes.size() >= INTRO_LENGTH_SIZE) {
+    length_in_sec = ConvertSPCStringToNum(length_bytes);
+    qLog(Debug) << length_in_sec << "------ LENGTH";
+
+    if (!length_in_sec || length_in_sec >= 0x1FFF) {
+      // This means that parsing the length as a string failed, so get value LE.
+      length_in_sec =
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:73]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+  }
+
+  file.seek(FADE_LENGTH_OFFSET);
+  QByteArray fade_bytes = file.read(FADE_LENGTH_SIZE);
+  if (fade_bytes.size() >= FADE_LENGTH_SIZE) {
+    quint64 fade_length_in_ms = ConvertSPCStringToNum(fade_bytes);
+    qLog(Debug) << fade_length_in_ms << "------ Fade Length";
+
+    if (fade_length_in_ms > 0x7FFF) {
+      fade_length_in_ms = fade_bytes[0] | (fade_bytes[1] << 8) |
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:89]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+  }
+
+  /*  Check for XID6 data -- this is infrequently used, but being able to fill
+   * in data from this is ideal before trying to rely on APETAG values. XID6
+   * format follows EA's binary file format standard named "IFF" */
+  file.seek(XID6_OFFSET);
+  if (has_id6 && file.read(4) == QString("xid6").toLatin1()) {
+    QByteArray xid6_head_data = file.read(4);
+    if (xid6_head_data.size() >= 4) {
+      qint64 xid6_size = xid6_head_data[0] | (xid6_head_data[1] << 8) |
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:101]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+  quint64 result = 0;
+  for (auto it = arr.begin(); it != arr.end(); it++) {
+    unsigned int num = *it - '0';
+    if (num > 9) break;
+    result = (result * 10) + num;  // Shift Left and add.
+  }
+  return result;
+}
+
+void GME::VGM::Read(const QFileInfo& file_info,
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:162]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+  quint64 length = 0;
+
+  if (!GetPlaybackLength(sample_count_bytes, loop_count_bytes, length)) return;
+
+  file.seek(GD3_TAG_PTR + pt);
+  QByteArray gd3_version = file.read(4);
+
+  file.seek(file.pos() + 4);
+  QByteArray gd3_length_bytes = file.read(4);
+  quint32 gd3_length =
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:192]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+  song_info->set_title(strings[0].toStdString());
+  song_info->set_album(strings[2].toStdString());
+  song_info->set_artist(strings[6].toStdString());
+  song_info->set_year(strings[8].left(4).toInt());
+  song_info->set_length_nanosec(length * kNsecPerMsec);
+  song_info->set_valid(true);
+  song_info->set_filetype(spb::tagreader::SongMetadata_FileType_VGM);
+}
+
+bool GME::VGM::GetPlaybackLength(const QByteArray& sample_count_bytes,
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:215]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+                              ((unsigned char)loop_count_bytes[1] << 8) |
+                              ((unsigned char)loop_count_bytes[2] << 16) |
+                              ((unsigned char)loop_count_bytes[3] << 24);
+
+  if (loop_sample_count <= 0) {
+    out_length = sample_count * 1000 / SAMPLE_TIMEBASE;
+    return true;
+  }
+
+  quint64 intro_length_ms =
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:241]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+    TagReaderGME::TagLib_Decode(tag->genre(), song_info->mutable_genre());
+  }
+
+  song_info->set_valid(true);
+  song_info->set_filetype(spb::tagreader::SongMetadata_FileType_SPC);
+}
+
+qint16 GME::SPC::GetNextMemAddressAlign32bit(qint16 input) {
+  return ((input + 0x3) & ~0x3);
+  // Plus 0x3 for rounding up (not down), AND NOT to flatten out on a 32 bit
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:148]: One line

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+                    spb::tagreader::SongMetadata* song_info) {
+  QFile file(file_info.filePath());
+  if (!file.open(QIODevice::ReadOnly)) return;
+
+  qLog(Debug) << "Reading tags from SPC file: " << file_info.fileName();
+
+  // Check for header -- more reliable than file name alone.
+  if (!file.read(33).startsWith(QString("SNES-SPC700").toLatin1())) return;
+
+  /*
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:43]: Usually use // comments except for the header

**Review by jonaski — CHANGES_REQUESTED** [review]
- Should have a copyright header
- spc and vgm should be added to src/widgets/fileview.cpp too

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 25]
```diff
@@ -0,0 +1,284 @@
...
+
+#include <apefile.h>
+#include <tag.h>
+
+#include <QByteArray>
+#include <QChar>
+#include <QFile>
+#include <QFileInfo>
+#include <QString>
+#include <QtEndian>
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:11]: You don't need QtEndian

File: ext/libstrawberry-tagreader/gmereader.h [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+
+}  // namespace GME
+
+/* 
+ * TagReaderGME
+ * Fulfills Strawberry's Intended interface for tag reading.
+ */
+class TagReaderGME : public TagReaderBase {
+
+public:
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.h:80]:
```suggestion
 public:
```

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+
+  file.seek(ARTIST_OFFSET);
+  song_info->set_artist(QString::fromLatin1(file.read(32)).toStdString());
+
+  file.seek(INTRO_LENGTH_OFFSET);
+  QByteArray length_bytes = file.read(INTRO_LENGTH_SIZE);
+  quint64 length_in_sec = 0;
+  if (length_bytes.size() >= INTRO_LENGTH_SIZE) {
+    length_in_sec = ConvertSPCStringToNum(length_bytes);
+    qLog(Debug) << length_in_sec << "------ LENGTH";
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:69]: Think we can remove this debug line

File: ext/libstrawberry-tagreader/gmereader.cpp (hunk 12 above)
Comment (Eoin-ONeill-Yokai) [reply ext/libstrawberry-tagreader/gmereader.cpp:212]: This should be addressed now with a compile time conditional.

**Comment by Eoin-ONeill-Yokai** [comment]
I had some time last weekend and made many of the changes requested. Let me know if there's anything else needed here.

**Comment by Eoin-ONeill-Yokai** [comment]
> I had some time last weekend and made many of the changes requested. Let me know if there's anything else needed here.

Hmmm -- Looks like there's still some building issues on 15.4 QT. Perhaps my conditional check is on the wrong version?

I'll have to correct that.

File: ext/libstrawberry-tagreader/gmereader.cpp [hunk 28]
```diff
@@ -0,0 +1,288 @@
...
+  QByteArray gd3_version = file.read(4);
+
+  file.seek(file.pos() + 4);
+  QByteArray gd3_length_bytes = file.read(4);
+  quint32 gd3_length = GME::UnpackBytes32(gd3_length_bytes, gd3_length_bytes.size());
+
+  QByteArray gd3Data = file.read(gd3_length);
+  QTextStream fileTagStream(gd3Data, QIODevice::ReadOnly);
+  // Stored as 16 bit UTF string, two bytes per letter.
+#if QT_VERSION > QT_VERSION_CHECK(5, 15, 0)
```
Comment (jonaski) [inline ext/libstrawberry-tagreader/gmereader.cpp:211]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

**Review by jonaski — CHANGES_REQUESTED** [review]

**Comment by jonaski** [comment]
I manually rebased and merged this in https://github.com/strawberrymusicplayer/strawberry/commit/80da565609fc84dea90e9e30b1b870263e49fc18
I made a couple of changes:
- I changed `TagReaderBase::ReadFile` to return `bool`, it returns true when the format is recognized by the reader.
- Since GME uses `TagLib::Tag` and `TagLib::APE::File`, it's included only when using TagLib.
- Some minor formatting changes.

**Comment by jonaski** [comment]
This doesn't look right, shouldn't it be UTF-16 also with Qt 5? That's what's used in Clementine.
```
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
  fileTagStream.setEncoding(QStringConverter::Utf16);
#else
  fileTagStream.setCodec("UTF-8");
#endif
```

**Comment by Eoin-ONeill-Yokai** [comment]
> This doesn't look right, shouldn't it be UTF-16 also with Qt 5? That's what's used in Clementine.
> 
> ```
> #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
>   fileTagStream.setEncoding(QStringConverter::Utf16);
> #else
>   fileTagStream.setCodec("UTF-8");
> #endif
> ```

Yeah my mistake. You're correct here.

**Pull request closed without merging**